Fix Devcontainer generator with --dev option path error

thejonroberts opened the pull request:
### Motivation / Background
Fixes #52704 

This Pull Request has been created because the devcontainer generator with the --dev option fails (`rails g devcontainer --dev`).  The app generator --dev works just fine.

### Detail

This Pull request moves the `Rails::Generators::RAILS_DEV_PATH` definition to Rails::Generators module rather than adding it to Rails::Generators in AppGenerator definition. It also creates a DEV_RAILS_PATH constant in `railties/test/generators/generators_test_helper.rb`, so that it can be used consistently in tests.

### Additional information

- I added a generator test file, it seems the generator is not tested for the `rails generate` usage, though it is well tested via app generator. I added tests for the dev option behavior only. I can work on full specs, but will keep it separate from this.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

jeromedalbert commented on railties/lib/rails/generators/rails/devcontainer/devcontainer_generator.rb:
```diff
@@ -140,10 +140,16 @@ def devcontainer_db_service_yaml(**options)
-            target: Rails::Generators::RAILS_DEV_PATH
+            source: rails_dev_path,
+            target: rails_dev_path
           }
```
`Rails::Generators::RAILS_DEV_PATH` is defined in `app_generator.rb`:

https://github.com/rails/rails/blob/c6e3336cfdec5ae9c7fb98b03d0196aa56ab30f9/railties/lib/rails/generators/rails/app/app_generator.rb#L285-L288

When executing the `devcontainer` generator alone, `app_generator.rb` is never required, so the generator errors out because it doesn't know about the constant. This is the root cause of the issue.

---

Maybe it would be better if the constant could be reused, which would also result in a shorter fix.

I suggest maybe moving the `RAILS_DEV_PATH` constant to `railties/lib/rails/generators.rb`, so both `devcontainer_generator.rb` and `app_generator.rb` have access to it.

Alternatively `require "rails/generators/rails/app/app_generator"` could be added at the top of `devcontainer_generator.rb` or at the beginning of `def local_rails_mount`, but that would result in circular requires because `app_generator.rb` also requires `devcontainer_generator.rb`. So this is probably not a good idea.

thejonroberts replied:
Thanks, Jerome!  I like the `railties/lib/rails/generators.rb` definition idea. Trying it out now and seems to work, besides breaking shared `SharedGeneratorTests#test_dev_option`, which uses `Rails.root`... I think that is wrong - as far as I can tell, that's the directory of the created rails app, not the gem directory... pushing those changes up shortly.

thejonroberts replied:
Yes, that's much better. updated.

thejonroberts commented on railties/test/generators/shared_generator_tests.rb:
```diff
@@ -362,7 +362,7 @@ def test_generator_when_sprockets_is_not_used
   def test_dev_option
     run_generator_using_prerelease [destination_root, "--dev"]
-    rails_path = File.expand_path("../../..", Rails.root)
+    rails_path = Rails::Generators::RAILS_DEV_PATH
```
Using constant in specs makes me want to test the constant... looking into it.

could do 
```rb
rails_path = File.expand_path("../../..", __dir__)
```
instead for these specs.

thejonroberts replied:
`Rails.root` & `destination_root` here represent the generated rails app directory. In these tests, that is `../test/fixtures/tmp`, so you can get to the "gem" root by going up from there.  But in actual usage, the generated app could be anywhere in relation to the local gem. But the right path for the rails gem **is always up the same level from here**. It's confusing, but I think __file__ or __dir__ communicate that a little better. We could even change `destination` for the generator and it should still pass.

jeromedalbert replied:
I see where you're coming from. After a quick search `destination_root` is used for going down, but for going up `__dir__` is used more.

Also found this constant which backs up the use of `__dir__` for going up: https://github.com/rails/rails/blob/c6e3336cfdec5ae9c7fb98b03d0196aa56ab30f9/railties/test/isolation/abstract_unit.rb#L29

(this constant is not accessible from your tests except if the file is required, which takes too long IMO)

thejonroberts replied:
Yes, I think a similar constant for generator tests makes a lot of sense. I will add something.

thejonroberts commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -283,10 +283,6 @@ def devcontainer
 
   module Generators
-    # We need to store the RAILS_DEV_PATH in a constant, otherwise the path
-    # can change in Ruby 1.8.7 when we FileUtils.cd.
```
Then I guess the constant is only to ensure consistency & avoid a lot of `File.expand_path` calls. Removing comment okay?

jeromedalbert replied:
I am not sure how to interpret the comment, it could mean that the path can change since Ruby 1.8.7. It is still true today that `FileUtils.cd` changes the working directory (which would affect `File.expand_path` among other things). I don't know what pre-1.8.7 behavior was. Maybe a conservative change could be to only remove "in Ruby 1.8.7" from the comment.

thejonroberts replied:
updated.

thejonroberts commented:
Tests for the rest of the generator functionality should probably be added to the new test file?

jeromedalbert commented:
The devcontainer generator functionality looks like it has been pretty extensively tested in [`app_generator_test.rb`](https://github.com/rails/rails/blob/c6e3336cfdec5ae9c7fb98b03d0196aa56ab30f9/railties/test/generators/app_generator_test.rb#L1242-L1461), since `AppGenerator` calls `DevcontainerGenerator` when the `--devcontainer` option is supplied. Are you suggesting that you want to move (or copy/share) those tests over to your new test file?

To me new test file looks good as is with just enough tests for some sanity checks and testing that the bug has been fixed. I am not well versed in the codebase's tests though.

jeromedalbert commented on railties/test/generators/devcontainer_generator_test.rb:
```diff
@@ -0,0 +1,42 @@
+
+          assert_equal "bind", mounts["type"]
+          assert_equal Rails::Generators::RAILS_DEV_PATH, mounts["source"]
+          assert_equal Rails::Generators::RAILS_DEV_PATH, mounts["target"]
```
Take my comments with a grain of salt as I am not familiar with the testing codebase and best practices, but it seems like using constants like `Rails::Generators::RAILS_DEV_PATH` inside tests relies on internal behavior which might not be very kosher. Replacing the constant with an path would allow testing only external behavior. Maybe some tests in [`app_generator_test.rb`](https://github.com/rails/rails/blob/main/railties/test/generators/app_generator_test.rb) could be helpful as a reference.

thejonroberts replied:
yeah, agreed, I think i saw it in an app generator test, but I don't like it, will change.

jeromedalbert commented on railties/test/generators/shared_generator_tests.rb:
```diff
@@ -362,7 +362,7 @@ def test_generator_when_sprockets_is_not_used
   def test_dev_option
     run_generator_using_prerelease [destination_root, "--dev"]
-    rails_path = File.expand_path("../../..", Rails.root)
+    rails_path = File.expand_path("../../..", __dir__)
     assert_file "Gemfile", %r{^gem ["']rails["'], path: ["']#{Regexp.escape rails_path}["']$}
```
```suggestion
    assert_file "Gemfile", %r{^gem ["']rails["'], path: ["']#{Regexp.escape RAILS_DEV_PATH}["']$}
```

thejonroberts replied:
Not available here, and I question whether it is worth requiring generators_test_helpers here. Already pretty clear that it is the rails dev path because of the gemfile path check.

jeromedalbert replied:
Ok no problem. By the way I don't seem to have the right to resolve comments, so if you could resolve this comment and the other outdated one that would result in an easier to read PR, thanks

jeromedalbert replied:
By the way, the constant would be available if it were defined above `module GeneratorsTestHelper` instead of inside it. Similar to how `RAILS_FRAMEWORK_ROOT` is defined.

thejonroberts replied:
OH... how about we just define it in `railties/test/abstract_unit` and follow the same pattern as the isolation tests! Using the same name! That works for all the tests. 

Seems so obvious now 🤣 Thanks for the help!

jeromedalbert replied:
Nice! Tests are failing right now though with a constant already initialized error

thejonroberts replied:
Yup! Turns out using the same constant name, in the global namespace, was not as brilliant as I thought!

I decided to use DEV_RAILS_PATH so it is noticeably different than the constant being tested, and leaving it in the GeneratorsTestHelper module, and not changing the shared test file where helpers aren't loaded.  

Could use the current global RAILS_FRAMEWORK_ROOT, but depends on require chains that I don't want to worry about!

thejonroberts committed:
Fix #52704 rails g devcontainer --dev error

Devcontainer generator errors with uninitialized constant
Rails::Generators::RAILS_DEV_PATH, which is defined with
AppGenerator in .../generators/rails/app/app_generator.rb.
Move definition to rails/generators.

Co-authored-by: Jerome Dalbert <jerome.dalbert@gmail.com>

andrewn617 commented:
This looks good to me. Thanks =)
